File: tests/ppapi_browser/ppp_instance/ppapi_ppp_instance.js
// Copyright (c) 2011 The Native Client Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

function startsWith(str, prefix) {
  return (str.indexOf(prefix) === 0);
}

function simulateMouseEvent(target, eventName) {
  var event = document.createEvent('MouseEvents');
  event.initMouseEvent(
      eventName,
      true,                       // bubbles
      true,                       // cancelable
      document.defaultView,
      1,                          // one click
      0, 0, 0, 0,                 // no co-ordinates
      false, false, false, false, // no key modifiers
      0,                          // button: 0 - left, 1 - middle, 2 - right
      null);                      // no related targets
  target.dispatchEvent(event);
}

function setupTests(tester, plugin) {
  //////////////////////////////////////////////////////////////////////////////
  // Test Helpers
  //////////////////////////////////////////////////////////////////////////////
  var numMessages = 0;
  function addTestListeners(numListeners, test, testFunction, runCheck) {
    var messageListener = test.wrap(function(message) {
      if (!startsWith(message.data, testFunction)) return;
      test.log(message.data);
      numMessages++;
      plugin.removeEventListener('message', messageListener, false);
      test.assertEqual(message.data, testFunction + ':PASSED');
      if (runCheck) test.assert(runCheck());
      if (numMessages < numListeners) {
        plugin.addEventListener('message', messageListener, false);
      } else {
        numMessages = 0;
        test.pass();
      }
    });
    plugin.addEventListener('message', messageListener, false);
  }

  function addTestListener(test, testFunction, runCheck) {
    return addTestListeners(1, test, testFunction, runCheck);
  }

  //////////////////////////////////////////////////////////////////////////////
  // Tests
  //////////////////////////////////////////////////////////////////////////////

  tester.addTest('PPP_Instance::DidCreate', function() {
    assertEqual(plugin.lastError, '');
  });

  tester.addAsyncTest('PPP_Instance::DidChangeView', function(test) {
    addTestListeners(3, test, 'DidChangeView');
  });

  tester.addAsyncTest('PPP_Instance::DidChangeFocus', function(test) {
    // TODO(polina): How can I simulate focusing on Windows?
    // For now just pass explicitely.
    if (startsWith(navigator.platform, 'Win')) {
      test.log('skipping test on ' + navigator.platform);
      test.pass();
      return;
    }
    addTestListeners(2, test, 'DidChangeFocus');
    plugin.tabIndex = 0;
    plugin.focus();
    plugin.blur();
  });

  tester.addAsyncTest('PPP_Instance::HandleInputEvent_Handled', function(test) {
    addTestListener(test, 'HandleInputEvent');
    // TODO(polina): why does this forward and bubble the event even though it
    // was handled? Based on headers, the handlers below should contain
    // test.fail(). Figure out if this is a bug (crbug.com/88728) or if the
    // header comments are off.
    var event = 'mousedown';
    plugin.addEventListener(
        event,
        function(e) { test.log('plugin received ' + event); },
        false);
    plugin.parentNode.addEventListener(
        event,
        function(e) { test.log('plugin.parentNode received ' + event); },
        false);
    simulateMouseEvent(plugin, event);
  });

  tester.addAsyncTest('PPP_Instance::HandleInputEvent_Bubbled', function(test) {
    var forwarded = false;
    var bubbled = false;
    var event = 'mouseup';
    addTestListener(test, 'HandleInputEvent',
                    function() { return forwarded && bubbled; });
    plugin.addEventListener(event,
                            function(e) { forwarded = true; }, false);
    plugin.parentNode.addEventListener(event,
                                       function(e) { bubbled = true; }, false);
    simulateMouseEvent(plugin, event);
  });

  // TODO(polina/bbudge/abarth):
  // PPP_Instance::HandleDocumentLoad is only used with full-frame plugins.
  // See if we can test this in a content handling extension test:
  // write an extension that contains a simple NEXE and then have an extension
  // API test that shows that, when you have that extension installed, the nexe
  // loads. There might be a way to load an extension in chrome using a
  // command-line parameter if you have an instance of chrome in your testing
  // harness.

  tester.addAsyncTest('PPP_Instance::DidDestroy', function(test) {
    // Destroy the plugin triggering PPP_Instance::DidDestroy.
    // A message should be logged to stdout.
    plugin.parentNode.removeChild(plugin);
    // Unfortunately, PPB_Messaging is a per-instance interface, so
    // posting messages from PPP_Instance::DidDestory does not work.
    // TODO(polina): Is there a different way to detect this from JavaScript?
    // For now just pass explicitely.
    test.pass();
  });
}

File: tests/ppapi_browser/ppp_instance/ppapi_ppp_instance.cc
// Copyright (c) 2011 The Native Client Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include <string.h>

#include <vector>

#include "native_client/src/include/nacl_macros.h"
#include "native_client/src/shared/platform/nacl_check.h"
#include "native_client/tests/ppapi_test_lib/get_browser_interface.h"
#include "native_client/tests/ppapi_test_lib/test_interface.h"

#include "ppapi/c/pp_bool.h"
#include "ppapi/c/pp_errors.h"
#include "ppapi/c/pp_input_event.h"
#include "ppapi/c/ppp_instance.h"

namespace {

PP_Bool DidCreate(PP_Instance instance,
                  uint32_t argc,
                  const char* argn[],
                  const char* argv[]) {
  printf("--- PPP_Instance::DidCreate\n");
  PP_Bool status = DidCreateDefault(instance, argc, argn, argv);
  // TEST_PASSED has no effect from this function.
  // But as long as the plugin loads and tests are run, we know this succeeded.
  // See ppapi_browser/bad for failing DidCreate.
  return status;
}

void DidDestroy(PP_Instance instance) {
  printf("--- PPP_Instance::DidDestroy\n");
  CHECK(instance == pp_instance());
  // TEST_PASSED has no effect from this function.
  // TODO(polina): how else can I test this programmatically?
  // One option is to use golden files. The other option is to crash the nexe
  // and rely on a status event when we implement crash detection in the
  // trusted plugin.
}

void DidChangeView(PP_Instance instance,
                   const struct PP_Rect* position,
                   const struct PP_Rect* clip) {
  printf("--- PPP_Instance::DidChangeView\n");
  EXPECT(instance == pp_instance());
  EXPECT(clip->point.x == 0 && clip->point.y == 0);
  // These are based on embed dimensions.
  EXPECT(position->size.width == 15 && clip->size.width == 15);
  EXPECT(position->size.height == 20 && clip->size.height == 20);

  TEST_PASSED;
}

void DidChangeFocus(PP_Instance instance,
                    PP_Bool has_focus) {
  printf("--- PPP_Instance::DidChangeFocus has_focus=%d\n", has_focus);
  // There should be no focus on load, so this will trigger when we gain it
  // and then release it and so on.
  static bool expected_has_focus = true;
  EXPECT(instance == pp_instance());
  EXPECT(has_focus == expected_has_focus);
  expected_has_focus = !expected_has_focus;

  TEST_PASSED;
}

PP_Bool HandleInputEvent(PP_Instance instance,
                         const struct PP_InputEvent* event) {
  printf("--- PPP_Instance::HandleInputEvent event=%d\n", event->type);
  EXPECT(instance == pp_instance());

  TEST_PASSED;

  if (event->type == PP_INPUTEVENT_TYPE_MOUSEDOWN)
    return PP_TRUE;  // Do not forward to the web page.
  else
    return PP_FALSE;  // Forward to the web page.
}

PP_Bool HandleDocumentLoad(PP_Instance instance,
                           PP_Resource url_loader) {
  NACL_NOTREACHED();  // Only called for full-frame plugins.
  return PP_FALSE;
}

const PPP_Instance ppp_instance_interface = {
  DidCreate,
  DidDestroy,
  DidChangeView,
  DidChangeFocus,
  HandleInputEvent,
  HandleDocumentLoad
};

}  // namespace


void SetupTests() {
  // Each PPP_Instance function called by the browser acts as a test.
}

void SetupPluginInterfaces() {
  RegisterPluginInterface(PPP_INSTANCE_INTERFACE, &ppp_instance_interface);
}
